File: weeks/7/3/solution.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

typedef long long ll;

constexpr int MAX_NUM_ISLANDS = int(1e4+1);

vector<int> bridges[MAX_NUM_ISLANDS];
bool visited[MAX_NUM_ISLANDS];

void solve() {
  int numIslands, numBridges, startIsland;
  cin >> numIslands >> numBridges >> startIsland;

  for (int i = 0; i < numBridges; ++i) {
    int source, dest;
    cin >> source >> dest;

    bridges[source].push_back(dest);
  }

  visited[startIsland] = true;

  queue<pair<int, int>> bfsQueue;
  bfsQueue.push({ startIsland, 0 }); // 섬, 거리

  while (!bfsQueue.empty()) {
    auto [island, distance] = bfsQueue.front();
    bfsQueue.pop();

    // 방문하지 않은 다음 섬은 큐에 대기
    for (int neighbor : bridges[island]) {
      if (neighbor == startIsland) { // 사이클 발견 시
        cout << distance+1 << '\n';
        return;
      }
      if (visited[neighbor]) {
        continue;
      }

      visited[neighbor] = true;
      bfsQueue.push({ neighbor, distance+1 });
    }
  }

  cout << -1 << '\n';
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  solve();
}

File: weeks/7/3/solution.py
import sys
sys.setrecursionlimit(10**6)
readline = sys.stdin.readline
from collections import defaultdict, deque

numIslands, numBridges, startIsland = map(int, readline().split())

bridges = defaultdict(list)
for _ in range(numBridges):
    source, dest = map(int, readline().split())

    bridges[source].append(dest)

visited = [False] * (numIslands+1)
visited[startIsland] = True

def walk():
    bfsQueue = deque()
    bfsQueue.append((startIsland, 0)) # 섬, 거리

    while len(bfsQueue) > 0:
        island, distance = bfsQueue.popleft()

        # 방문하지 않은 다음 섬은 큐에 대기
        for nextIsland in bridges[island]:
            if nextIsland == startIsland: # 사이클 발견 시
                return distance+1
            if visited[nextIsland]:
                continue

            visited[nextIsland] = True
            bfsQueue.append((nextIsland, distance+1))

    return -1

print(walk())
